### ISSUE TITLE
Cannot switch module from forge to git

### ISSUE THREAD

**Issue by adrienthebo** [puppetlabs/r10k#30]
The Forge and Git modules don't take into consideration that an existing module might be switched from one to the other. As such, doing this could make things explode pretty badly. Modules should be able to determine if they need to wipe an existing directory when installing or updating.

This issue addresses the problem raised in https://github.com/adrienthebo/r10k/issues/2#issuecomment-14225805

**Comment by eliasp** [comment]
A related issue is: replacing a modules implementation, e.g. changing this:

```
mod 'puppetlabs/firewall'
```

to this:

```
mod 'firewall'
    :git => 'https://github.com/example42/puppet-firewall.git'
```

The result is this error message:

```
root@puppetmaster production [production] # r10k puppetfile install
[R10K::TaskRunner - ERROR] Task #<R10K::Task::Module::Sync:0x00000002884c60> failed while running:
"git clone --reference /root/.r10k/git/https---github.com-example42-puppet-firewall.git
https://github.com/example42/puppet-firewall.git /etc/puppet/environments/production/modules/firewall"
returned with non-zero exit value 128
```

**Comment by ssm** [comment]
Just to add the git error message to this ticket.

The error message from r10k is:

```
[R10K::TaskRunner - ERROR] Task #<R10K::Task::Module::Sync:0x000000020a6020> failed while running: \
  "git clone --reference /srv/puppet/cache/git---github.com-puppetlabs-puppetlabs-apt.git \
  git://github.com/puppetlabs/puppetlabs-apt.git /srv/puppet/env/production/modules/apt" \
  returned with non-zero exit value 128
```

While the error message from that git command line is:

```
# git clone --reference /srv/puppet/cache/git---github.com-puppetlabs-puppetlabs-apt.git \
  git://github.com/puppetlabs/puppetlabs-apt.git /srv/puppet/env/production/modules/apt
fatal: destination path '/srv/puppet/env/production/modules/apt' already exists \
 and is not an empty directory.
```

Deleting the directory /srv/puppet/env/production/modules/apt (and for the other environments as well) lets r10k check out the module from the cache git repository.

**Comment by adrienthebo** [comment]
This issue should be fixed in 1.2.0, if this still persists reopen this ticket and we can continue this discussion.
